blockcerts/certifiedblock.go

```diff
@@ -0,0 +1,117 @@
+    "github.com/spacemeshos/go-spacemesh/hare"
+    "github.com/spacemeshos/go-spacemesh/log"
+    "github.com/spacemeshos/go-spacemesh/p2p"
+    "github.com/spacemeshos/go-spacemesh/p2p/pubsub"
+    "github.com/spacemeshos/go-spacemesh/signing"
+    "github.com/spacemeshos/go-spacemesh/sql"
+    "math"
+    "sync"
+)
+
+const (
+    blockCertifierRole uint32 = math.MaxUint32 - 2        // for Rolacle
+    BlockSigTopic             = "trackableBlockSignature" // for gossip
+
+)
+
+// BlockCertifyingService is a long-lived service responsible for participating
+// in block-signing upon hare termination.
+// Inputs: hareTerminationsCh and GossipHandler.
+// Outputs: gossipPublisher and database
+type BlockCertifyingService struct {
+    // Inputs & Outputs
+    hareTerminationsCh <-chan hare.TerminationBlockOutput
+    gossipPublisher    pubsub.Publisher
+    db                 sql.Executor
+    // Dependencies
+    rolacle     hare.Rolacle
+    config      certtypes.BlockCertificateConfig
+    blockSigner signing.Signer
+    // Internal
+    signatureCache   sigCache
+    completedCertsCh chan certtypes.BlockCertificate
+
+    logger log.Logger
+}
+
+// NewBlockCertifyingService constructs a new BlockCertifyingService.
+func NewBlockCertifyingService(
+    hareTerminations <-chan hare.TerminationBlockOutput,
+    rolacle hare.Rolacle,
+    gossipPublisher pubsub.Publisher,
+    blockSigner signing.Signer,
+    db sql.Executor,
+    config certtypes.BlockCertificateConfig,
+    logger log.Logger,
+) (*BlockCertifyingService, error) {
+    service := &BlockCertifyingService{}
+    service.rolacle = rolacle
+    service.config = config
+    service.hareTerminationsCh = hareTerminations
+    service.gossipPublisher = gossipPublisher
+    service.blockSigner = blockSigner
+    service.db = db
+
+    service.signatureCache = sigCache{
+        logger:             logger,
+        blockSigsByLayer:   sync.Map{},
+        cacheBoundary:      types.NewLayerID(0), // TODO: check this is okay
+        cacheBoundaryMutex: sync.RWMutex{},
+        completedCertsCh:   service.completedCertsCh,
+    }
+    service.completedCertsCh = make(chan certtypes.BlockCertificate, 100)
+    // TODO: empirically decide on adequate channel size
+
+    service.logger = logger
+    return service, nil
+}
+
+func (s *BlockCertifyingService) Start(ctx context.Context) error {
+    go blockSigningLoop(ctx, s.hareTerminationsCh,
+        s.blockSigner, s.config.CommitteeSize, s.rolacle,
+        s.gossipPublisher, &s.signatureCache, s.logger,
+    )
+    go certificateStoringLoop(ctx, s.completedCertsCh, s.db, s.logger)
+    return nil
+}
+
+func (s *BlockCertifyingService) SignatureCacher() SigCacher {
+    return &s.signatureCache // exported functions are thread safe
+}
+
+// GossipHandler returns a function that handles incoming BlockSignatureMsg gossip.
+func (s *BlockCertifyingService) GossipHandler() pubsub.GossipHandler {
```

Comment:
i don't get why to create a clojure from this method

`GossipHandler(ctx context.Context, peerID p2p.Peer, bytes []byte)`

Reply:
So that if we want to change where the Gossip Handler is defined in the future (or we want to switch it out depending on some state of the service or add extra "call only once logic") we don't need to break the exported api of the BlockCertifyingService. Plenty of good reasons, I think?
It decouples the handler implementation from the BlockCertifyingService exported API.

Reply:
please don't code for the future. i call this premature abstraction.
rather, code consistently with the rest of the codebase.

if you are convinced that we need it, change all the gossip handlers in a separate PR.